File: languages/csharp/AvalonSdk/CrossNodeLogin.cs
// Cross-node login — mirrors crates/sdk/src/cross_node_login.rs,
// itself mirroring device_login's own start-then-poll shape almost exactly: from a caller's
// perspective both flows look identical (POST .../start on this node, show the user a code,
// poll until approved, get back a real Session). The one structural difference: POST
// /auth/cross-node/start returns no verification_uri the way device pairing's would — there
// is no Hub route for cross-node approval yet, just a UserCode and
// RequestingContext to show however this integrator's own UI displays a pairing code.
//
// Also exposes the same-device fast path (AvalonClient.SubmitCrossNodeLoginGrantAsync):
// epic #623's own scope note draws a real distinction between the cross-device dance above
// (needed when the approving human is somewhere that genuinely doesn't hold the identity's
// signing key) and a caller that already does — not the common case for this SDK's usual
// integrator-backend callers (who never hold a *player's* own key), but a real one for a
// deployment that directly controls some identity's key material (e.g. a service/bot
// identity). That caller can mint, sign, and submit a grant in one call, skipping the
// start/poll dance entirely.

using System;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading;
using System.Threading.Tasks;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.Crypto.Signers;

namespace Avalon.Sdk
{
    /// <summary>Mirrors <c>SdkError::CrossNodeLoginDenied</c> — the request was explicitly
    /// denied (POST /auth/cross-node/deny).</summary>
    public sealed class CrossNodeLoginDeniedException : Exception
    {
        public CrossNodeLoginDeniedException() : base("cross-node login was denied")
        {
        }
    }

    /// <summary>Mirrors <c>SdkError::CrossNodeLoginExpired</c> — the request's ~10-minute TTL
    /// (<c>crates/server/src/cross_node_login.rs</c>) elapsed before it was approved or
    /// denied.</summary>
    public sealed class CrossNodeLoginExpiredException : Exception
    {
        public CrossNodeLoginExpiredException() : base("cross-node login expired before it was approved")
        {
        }
    }

    /// <summary>
    /// A pending cross-node login, returned by
    /// <see cref="AvalonClient.CrossNodeLoginAsync"/>. Mirrors the Rust SDK's
    /// <c>cross_node_login::CrossNodeLogin</c>.
    /// </summary>
    public sealed class CrossNodeLogin
    {
        /// <summary>Same cap the Rust SDK's own <c>MAX_POLL_INTERVAL_SECONDS</c> documents.</summary>
        private const long MaxPollIntervalSeconds = 60;

        private readonly AvalonClient _client;
        private readonly string _requestCode;
        private long _pollIntervalSeconds;

        internal CrossNodeLogin(
            AvalonClient client,
            string requestCode,
            string userCode,
            string requestingContext,
            long expiresIn,
            long pollIntervalSeconds)
        {
            _client = client;
            _requestCode = requestCode;
            UserCode = userCode;
            RequestingContext = requestingContext;
            ExpiresIn = expiresIn;
            // Same floor the Rust SDK's own DEFAULT_POLL_INTERVAL_SECONDS documents — a server
            // response should always carry its own poll_interval, this is only a fallback.
            _pollIntervalSeconds = pollIntervalSeconds > 0 ? pollIntervalSeconds : 5;
        }

        /// <summary>Short, human-typeable code — show it however this integrator's own UI
        /// displays a pairing code. Unlike a device-pairing verification URI, there's no
        /// ready-made Hub route to point at yet (#639, not built).</summary>
        public string UserCode { get; }

        /// <summary>The requesting node's own context, meant to be shown by whatever eventually
        /// approves this (Hub/mobile-hub, once built) so a human can tell what they're approving
        /// login to.</summary>
        public string RequestingContext { get; }

        /// <summary>Seconds until this request expires if it's never approved.</summary>
        public long ExpiresIn { get; }

        /// <summary>
        /// Drives POST /auth/cross-node/poll to completion — identical backoff shape to the
        /// Rust SDK's <c>CrossNodeLogin::wait</c> (doubling the interval, capped at
        /// <see cref="MaxPollIntervalSeconds"/>, on <c>slow_down</c> rather than treating it as
        /// plain <c>pending</c>). Resolves to a real <see cref="Session"/> on <c>approved</c> by
        /// feeding the minted token through <see cref="AvalonClient.AuthenticateAsync"/> — the
        /// same path a normal login already uses — or throws
        /// <see cref="CrossNodeLoginDeniedException"/>/<see cref="CrossNodeLoginExpiredException"/>
        /// on <c>denied</c>/<c>expired</c>.
        /// </summary>
        public async Task<Session> WaitAsync(CancellationToken ct = default)
        {
            var interval = Math.Max(_pollIntervalSeconds, 1);

            while (true)
            {
                await Task.Delay(TimeSpan.FromSeconds(interval), ct).ConfigureAwait(false);

                using var request = new HttpRequestMessage(HttpMethod.Post, $"{_client.ServerUrl}/auth/cross-node/poll");
                request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _requestCode);
                using var response = await _client.Http.SendAsync(request, ct).ConfigureAwait(false);
                if (!response.IsSuccessStatusCode)
                {
                    throw Session.ServerError(response.StatusCode);
                }
                var body = await Session.ReadJsonAsync<Avalon.Sdk.Generated.PollCrossNodeLoginResponse>(response, ct).ConfigureAwait(false);

                switch (body.Status)
                {
                    case "pending":
                        continue;
                    case "slow_down":
                        interval = Math.Min(interval * 2, MaxPollIntervalSeconds);
                        continue;
                    case "denied":
                        throw new CrossNodeLoginDeniedException();
                    case "approved":
                        // Single-use delivery (cross_node_login.rs's own invariant): a missing
                        // token here would mean this response already lost a consumption race,
                        // which the server surfaces as expired, not approved with no token —
                        // treat it the same as a malformed response.
                        if (body.Token is null)
                        {
                            throw Session.ServerError(System.Net.HttpStatusCode.InternalServerError);
                        }
                        return await _client.AuthenticateAsync(body.Token, ct).ConfigureAwait(false);
                    default:
                        // "expired" and anything unrecognized both mean this request is done and
                        // will never resolve to a session.
                        throw new CrossNodeLoginExpiredException();
                }
            }
        }
    }

    public sealed partial class AvalonClient
    {
        /// <summary>Starts a cross-node login via POST /auth/cross-node/start
        /// against this client's own <see cref="AvalonConfig.ServerUrl"/> — the node being
        /// logged into, which may not be the identity's "home" node at all.</summary>
        public async Task<CrossNodeLogin> CrossNodeLoginAsync(CancellationToken ct = default)
        {
            // No idempotency key: a retried start would just mint a second, independent request
            // rather than replay the first, same reasoning as every other unkeyed write.
            using var request = new HttpRequestMessage(HttpMethod.Post, $"{ServerUrl}/auth/cross-node/start");
            using var response = await Http.SendAsync(request, ct).ConfigureAwait(false);
            if (!response.IsSuccessStatusCode)
            {
                throw Session.ServerError(response.StatusCode);
            }
            var body = await Session.ReadJsonAsync<Avalon.Sdk.Generated.StartCrossNodeLoginResponse>(response, ct).ConfigureAwait(false);

            return new CrossNodeLogin(
                this,
                body.RequestCode,
                body.UserCode,
                body.RequestingContext,
                body.ExpiresIn,
                body.PollInterval);
        }

        /// <summary>The exact bytes a <c>CrossNodeLoginGrant</c>'s signature covers — must match
        /// <c>avalon_protocol::cross_node_login::signing_bytes</c> exactly. This SDK defines its
        /// own copy rather than depending on the server's private construction, same posture as
        /// every other signed request in this repo (see <c>Achievements.cs</c>'s own
        /// <c>AttestationSigningBytes</c>).</summary>
        private static byte[] CrossNodeLoginSigningBytes(
            Guid identityId,
            Guid signingKeyId,
            string destinationBaseUrl,
            string requestingContext,
            Guid nonce,
            DateTimeOffset issuedAt,
            DateTimeOffset expiresAt)
        {
            var text =
                $"avalon:cross-node-login:v1:{identityId}:{signingKeyId}:{destinationBaseUrl}:" +
                $"{requestingContext}:{nonce}:{issuedAt.ToUnixTimeSeconds()}:{expiresAt.ToUnixTimeSeconds()}";
            return Encoding.UTF8.GetBytes(text);
        }

        /// <summary>Pure-managed Ed25519 via BouncyCastle, same library (and Unity/IL2CPP-safe
        /// reasoning) <c>Achievements.cs</c>'s own <c>SignWithIssuerKey</c> already
        /// establishes — this is the identity-level counterpart, signing with the *player's*
        /// own event-signing key seed rather than an integrator's issuer key.</summary>
        private static byte[] SignWithIdentityKey(byte[] signingKeySeed, byte[] message)
        {
            var privateKey = new Ed25519PrivateKeyParameters(signingKeySeed, 0);
            var signer = new Ed25519Signer();
            signer.Init(true, privateKey);
            signer.BlockUpdate(message, 0, message.Length);
            return signer.GenerateSignature();
        }

        private static string ToLowerHex(byte[] bytes)
        {
            var sb = new StringBuilder(bytes.Length * 2);
            foreach (var b in bytes)
            {
                sb.Append(b.ToString("x2"));
            }
            return sb.ToString();
        }

        /// <summary>
        /// The same-device fast path (see this file's header comment): mints, signs, and
        /// submits a <c>CrossNodeLoginGrant</c> directly against this client's own
        /// <see cref="AvalonConfig.ServerUrl"/>, for a caller that already holds
        /// <paramref name="identityId"/>'s real 32-byte Ed25519 event-signing key seed
        /// (<paramref name="signingKeySeed"/>, identified by <paramref name="signingKeyId"/>) —
        /// skipping <see cref="CrossNodeLoginAsync"/>/<see cref="CrossNodeLogin.WaitAsync"/>
        /// entirely. Resolves to a real <see cref="Session"/> the same way <c>WaitAsync</c>
        /// does, via <see cref="AuthenticateAsync"/>.
        /// </summary>
        public async Task<Session> SubmitCrossNodeLoginGrantAsync(
            Guid identityId,
            Guid signingKeyId,
            byte[] signingKeySeed,
            CancellationToken ct = default)
        {
            var issuedAt = DateTimeOffset.UtcNow;
            // Same TTL avalon_protocol::cross_node_login::DEFAULT_TTL_SECONDS documents — this
            // SDK defines its own copy for the same reason CrossNodeLoginSigningBytes does.
            var expiresAt = issuedAt.AddSeconds(60);
            var nonce = Guid.NewGuid();
            var destinationBaseUrl = ServerUrl;
            // No Hub-driven context yet to show a human — the destination itself is the
            // most honest context this caller can supply on its own behalf.
            var requestingContext = destinationBaseUrl;

            var signingBytes = CrossNodeLoginSigningBytes(
                identityId, signingKeyId, destinationBaseUrl, requestingContext, nonce, issuedAt, expiresAt);
            var signature = SignWithIdentityKey(signingKeySeed, signingBytes);

            var grant = new Avalon.Sdk.Generated.CrossNodeLoginGrant
            {
                IdentityId = identityId,
                SigningKeyId = signingKeyId,
                DestinationBaseUrl = destinationBaseUrl,
                RequestingContext = requestingContext,
                Nonce = nonce,
                IssuedAt = issuedAt,
                ExpiresAt = expiresAt,
                Signature = ToLowerHex(signature),
            };

            using var request = new HttpRequestMessage(HttpMethod.Post, $"{ServerUrl}/auth/cross-node/submit");
            request.Content = new StringContent(
                JsonSerializer.Serialize(new Avalon.Sdk.Generated.SubmitGrantRequest { UserCode = null, Grant = grant }),
                Encoding.UTF8, "application/json");
            using var response = await Http.SendAsync(request, ct).ConfigureAwait(false);
            if (!response.IsSuccessStatusCode)
            {
                throw Session.ServerError(response.StatusCode);
            }
            var body = await Session.ReadJsonAsync<Avalon.Sdk.Generated.SubmitGrantResponse>(response, ct).ConfigureAwait(false);
            if (body.Token is null)
            {
                throw Session.ServerError(System.Net.HttpStatusCode.InternalServerError);
            }
            return await AuthenticateAsync(body.Token, ct).ConfigureAwait(false);
        }

        // --- The approving device's own half ---
        //
        // CrossNodeLoginAsync/WaitAsync/SubmitCrossNodeLoginGrantAsync above are
        // the *requester's* side (the node being logged into). LookupCrossNodeLoginAsync/
        // DenyCrossNodeLoginAsync are the *approver's* side — whatever surface shows a human
        // the pairing code (no Hub route for this yet) looks the request up by its
        // short user code and can explicitly deny it. Neither call carries a bearer token:
        // the user code itself is the only credential, same shape device pairing's own
        // deny/resolve endpoints already use.

        /// <summary>GET /auth/cross-node/lookup?user_code=… — resolves a short pairing code
        /// into the full pending request (status, requesting context, whether the requesting
        /// node is a verified integrator) for an approval screen to render. Public — the user
        /// code itself is the only proof needed to look up its own request.</summary>
        public async Task<Avalon.Sdk.Generated.LookupCrossNodeLoginResponse> LookupCrossNodeLoginAsync(
            string userCode, CancellationToken ct = default)
        {
            var url = $"{ServerUrl}/auth/cross-node/lookup?user_code={Uri.EscapeDataString(userCode)}";
            using var request = new HttpRequestMessage(HttpMethod.Get, url);
            using var response = await Http.SendAsync(request, ct).ConfigureAwait(false);
            if (!response.IsSuccessStatusCode)
            {
                throw Session.ServerError(response.StatusCode);
            }
            return await Session.ReadJsonAsync<Avalon.Sdk.Generated.LookupCrossNodeLoginResponse>(response, ct).ConfigureAwait(false);
        }

        /// <summary>POST /auth/cross-node/deny — explicitly denies the pending request named
        /// by <paramref name="userCode"/>; <see cref="CrossNodeLogin.WaitAsync"/> on the
        /// requesting side then surfaces this as <see cref="CrossNodeLoginDeniedException"/>.
        /// Only a currently-<c>pending</c>, unexpired request can be denied.</summary>
        public async Task DenyCrossNodeLoginAsync(string userCode, CancellationToken ct = default)
        {
            using var request = new HttpRequestMessage(HttpMethod.Post, $"{ServerUrl}/auth/cross-node/deny");
            request.Content = new StringContent(
                JsonSerializer.Serialize(new Avalon.Sdk.Generated.UserCodeRequest { UserCode = userCode }),
                Encoding.UTF8, "application/json");
            using var response = await Http.SendAsync(request, ct).ConfigureAwait(false);
            if (!response.IsSuccessStatusCode)
            {
                throw Session.ServerError(response.StatusCode);
            }
        }
    }
}

File: languages/rust/src/cross_node_login.rs
//! `AvalonClient::cross_node_login()` (epic #623, issue #637) — the
//! ergonomic wrapper around #634's cross-node login lifecycle
//! (`crates/server/src/cross_node_login.rs`), mirroring `device_login`'s
//! own start-then-poll shape almost exactly: from a caller's perspective
//! both flows look identical (`POST .../start` on this node, show the user
//! a code, poll until approved, get back a real [`Session`]). The one
//! structural difference: `POST /auth/cross-node/start` returns no
//! `verification_uri` the way `device_login::DeviceLogin` does — there is
//! no Hub route for cross-node approval yet (#639, not built) — just a
//! `user_code` and `requesting_context` to show however this integrator's
//! own UI displays a pairing code.
//!
//! **Same-device fast path** ([`AvalonClient::submit_cross_node_login_grant`]):
//! epic #623's own scope note draws a real distinction between the
//! cross-device dance above (needed when the approving human is somewhere
//! that genuinely doesn't hold the identity's signing key) and a caller
//! that already does — which is not the common case for this SDK's usual
//! integrator-backend callers (who never hold a *player's* own key), but a
//! real one for a deployment that directly controls some identity's key
//! material (e.g. a service/bot identity). That caller can mint, sign, and
//! submit a grant in one call, skipping the start/poll dance entirely.

use ed25519_dalek::{Signer, SigningKey};
use serde::{Deserialize, Serialize};
use time::{Duration, OffsetDateTime};
use uuid::Uuid;

use crate::{AvalonClient, SdkError, Session};

/// How long a grant is valid for after `issued_at`, at mint time — the
/// verifying node independently re-checks `expires_at` itself. Short: a
/// grant is a one-shot approval consumed once by
/// `POST /auth/cross-node/submit`, not a re-presented claim.
pub const DEFAULT_TTL_SECONDS: i64 = 60;

/// A self-signed assertion that a human, shown `requesting_context`, just
/// approved logging `identity_id` (acting through `signing_key_id`) into
/// `destination_base_url`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct CrossNodeLoginGrant {
    /// The identity being logged in.
    pub identity_id: Uuid,
    /// Which of the identity's (possibly several) signing keys approved
    /// this.
    pub signing_key_id: Uuid,
    /// The node this grant is good for logging into, and nowhere else —
    /// inside the signed bytes, so an approved grant can't be relayed to a
    /// node the human never saw.
    pub destination_base_url: String,
    /// The human-legible context shown to the approver before they
    /// approved — carried in the signed bytes so the grant itself is
    /// evidence of what was shown, not just a claim about it.
    pub requesting_context: String,
    /// Anti-replay: unique per grant, checked against a consumed-nonce
    /// table by the verifying node.
    pub nonce: Uuid,
    /// When this grant was minted.
    #[serde(with = "time::serde::rfc3339")]
    pub issued_at: OffsetDateTime,
    /// When it stops being accepted.
    #[serde(with = "time::serde::rfc3339")]
    pub expires_at: OffsetDateTime,
    /// Lowercase hex-encoded Ed25519 signature over [`signing_bytes`].
    pub signature: String,
}

impl CrossNodeLoginGrant {
    /// The bytes this grant's own `signature` field covers — what both the
    /// approving client and the verifying node compute independently.
    pub fn signing_bytes(&self) -> Vec<u8> {
        signing_bytes(
            self.identity_id,
            self.signing_key_id,
            &self.destination_base_url,
            &self.requesting_context,
            self.nonce,
            self.issued_at,
            self.expires_at,
        )
    }
}

/// The exact bytes a [`CrossNodeLoginGrant`]'s signature covers —
/// deliberately excludes `signature` itself and includes every other field,
/// so a signature can never be replayed against a different identity, key,
/// destination, context, or validity window than the one it was actually
/// produced for.
///
/// Must stay byte-for-byte identical to the verifying node's own
/// construction (`avalon_protocol::cross_node_login::signing_bytes`) and to
/// the C#/TypeScript SDKs' — `conformance/vectors/cross-node-login.json`
/// is what proves it still is.
pub fn signing_bytes(
    identity_id: Uuid,
    signing_key_id: Uuid,
    destination_base_url: &str,
    requesting_context: &str,
    nonce: Uuid,
    issued_at: OffsetDateTime,
    expires_at: OffsetDateTime,
) -> Vec<u8> {
    format!(
        "avalon:cross-node-login:v1:{identity_id}:{signing_key_id}:{destination_base_url}:{requesting_context}:{nonce}:{}:{}",
        issued_at.unix_timestamp(),
        expires_at.unix_timestamp(),
    )
    .into_bytes()
}

/// Same floor `device_login`'s own `DEFAULT_POLL_INTERVAL_SECONDS`
/// documents, for the same reason — a server response should always carry
/// its own `poll_interval`, this is only a fallback.
const DEFAULT_POLL_INTERVAL_SECONDS: i64 = 5;
/// Same cap `device_login::MAX_POLL_INTERVAL_SECONDS` documents.
const MAX_POLL_INTERVAL_SECONDS: i64 = 60;

#[derive(Deserialize)]
struct StartCrossNodeLoginResponse {
    request_code: String,
    user_code: String,
    requesting_context: String,
    expires_in: i64,
    poll_interval: i64,
}

#[derive(Deserialize)]
struct PollCrossNodeLoginResponse {
    status: String,
    token: Option<String>,
}

/// The approver-side read backing an approval screen — mirrors
/// `crates/server/src/cross_node_login.rs::LookupCrossNodeLoginResponse`
/// at the wire level. Never carries `request_code` (the polling device's
/// own bearer credential, not the approver's business).
#[derive(Debug, Clone, Deserialize)]
pub struct CrossNodeLoginLookup {
    /// One of `"pending"`, `"denied"`, `"expired"`, `"approved"` — an
    /// approval screen only ever meaningfully acts on `"pending"`.
    pub status: String,
    /// The requesting node's own context to show a human before they
    /// decide whether to approve.
    pub requesting_context: String,
    /// Seconds until this request expires if it's never approved.
    pub expires_in: i64,
    /// Whether this node (the one being logged into) resolves to a real,
    /// registered integrator or known network anchor — issue #649,
    /// implementing #642's decided phishing-context requirement. Never a
    /// hard gate; an unverified requester still gets a prompt, just a
    /// clearly flagged one.
    pub integrator_verified: bool,
    /// A real registered display name — only present when
    /// `integrator_verified` is `true`.
    #[serde(default)]
    pub display_name: Option<String>,
}

/// [`AvalonClient::deny_cross_node_login`]'s response — mirrors
/// `crates/server/src/cross_node_login.rs::DenyResponse`.
#[derive(Debug, Clone, Deserialize)]
pub struct CrossNodeLoginDenial {
    /// Always `"denied"` on success.
    pub status: String,
}

#[derive(Serialize)]
struct SubmitGrantRequest<'a> {
    #[serde(skip_serializing_if = "Option::is_none")]
    user_code: Option<&'a str>,
    grant: &'a CrossNodeLoginGrant,
}

#[derive(Deserialize)]
struct SubmitGrantResponse {
    token: Option<String>,
}

/// A pending cross-node login (epic #623), returned by
/// [`AvalonClient::cross_node_login`]. Borrows the `AvalonClient` it was
/// started from, same shape `device_login::DeviceLogin` already
/// establishes.
pub struct CrossNodeLogin<'a> {
    client: &'a AvalonClient,
    request_code: String,
    /// Short, human-typeable code — show it however this integrator's own
    /// UI displays a pairing code. Unlike
    /// `device_login::DeviceLogin::verification_uri`, there's no
    /// ready-made Hub route to point at yet (#639, not built).
    pub user_code: String,
    /// The requesting node's own context, meant to be shown by whatever
    /// eventually approves this (Hub/mobile-hub, once built) so a human
    /// can tell what they're approving login to.
    pub requesting_context: String,
    /// Seconds until this request expires if it's never approved.
    pub expires_in: i64,
    poll_interval: i64,
}

impl AvalonClient {
    /// Starts a cross-node login (epic #623) via `POST /auth/cross-node/start`
    /// against this client's own `server_url` — the node being logged
    /// into, which may not be the identity's "home" node at all.
    pub async fn cross_node_login(&self) -> Result<CrossNodeLogin<'_>, SdkError> {
        // No idempotency key: same reasoning `device_login::login`'s own
        // doc comment gives — a retried start mints a second, independent
        // request rather than replaying the first.
        let response = crate::http::send(&self.http, &self.config.retry, false, |c| {
            c.post(format!("{}/auth/cross-node/start", self.config.server_url))
        })
        .await?;
        if !response.status().is_success() {
            return Err(crate::http::map_error_response(response).await);
        }
        let body: StartCrossNodeLoginResponse = response
            .json()
            .await
            .map_err(|e| SdkError::Protocol(e.to_string()))?;

        Ok(CrossNodeLogin {
            client: self,
            request_code: body.request_code,
            user_code: body.user_code,
            requesting_context: body.requesting_context,
            expires_in: body.expires_in,
            poll_interval: if body.poll_interval > 0 {
                body.poll_interval
            } else {
                DEFAULT_POLL_INTERVAL_SECONDS
            },
        })
    }

    /// The same-device fast path (see module doc comment): mints, signs,
    /// and submits a `CrossNodeLoginGrant` directly against this client's
    /// own `server_url`, for a caller that already holds `identity_id`'s
    /// real Ed25519 event-signing key (`signing_key`, identified by
    /// `signing_key_id`) — skipping `cross_node_login`/`CrossNodeLogin::wait`
    /// entirely. Resolves to a real [`Session`] the same way `wait` does,
    /// via [`AvalonClient::authenticate`].
    pub async fn submit_cross_node_login_grant(
        &self,
        identity_id: Uuid,
        signing_key_id: Uuid,
        signing_key: &SigningKey,
    ) -> Result<Session, SdkError> {
        let issued_at = OffsetDateTime::now_utc();
        let expires_at = issued_at + Duration::seconds(DEFAULT_TTL_SECONDS);
        let nonce = Uuid::new_v4();
        let destination_base_url = self.config.server_url.clone();
        // No Hub-driven context yet to show a human (#639) — the
        // destination itself is the most honest context this caller can
        // supply on its own behalf.
        let requesting_context = destination_base_url.clone();

        let bytes = signing_bytes(
            identity_id,
            signing_key_id,
            &destination_base_url,
            &requesting_context,
            nonce,
            issued_at,
            expires_at,
        );
        let signature = signing_key.sign(&bytes);
        let grant = CrossNodeLoginGrant {
            identity_id,
            signing_key_id,
            destination_base_url,
            requesting_context,
            nonce,
            issued_at,
            expires_at,
            signature: hex::encode(signature.to_bytes()),
        };

        let request = SubmitGrantRequest {
            user_code: None,
            grant: &grant,
        };
        let response = crate::http::send(&self.http, &self.config.retry, false, |c| {
            c.post(format!("{}/auth/cross-node/submit", self.config.server_url))
                .json(&request)
        })
        .await?;
        if !response.status().is_success() {
            return Err(crate::http::map_error_response(response).await);
        }
        let body: SubmitGrantResponse = response
            .json()
            .await
            .map_err(|e| SdkError::Protocol(e.to_string()))?;
        let token = body.token.ok_or_else(|| {
            SdkError::Protocol("same-device submit response was missing a token".to_string())
        })?;
        self.authenticate(&token).await
    }

    /// `GET /auth/cross-node/lookup?user_code=...` (closed out by
    /// #741/#749) — the approver's own half of the flow: reads real
    /// context for a `user_code` *before* deciding whether to approve or
    /// [`AvalonClient::deny_cross_node_login`] it. Unauthenticated —
    /// there's routinely no session yet on the node an approver reaches
    /// this from.
    pub async fn lookup_cross_node_login(
        &self,
        user_code: &str,
    ) -> Result<CrossNodeLoginLookup, SdkError> {
        let response = crate::http::send(&self.http, &self.config.retry, true, |c| {
            c.get(format!("{}/auth/cross-node/lookup", self.config.server_url))
                .query(&[("user_code", user_code)])
        })
        .await?;
        if !response.status().is_success() {
            return Err(crate::http::map_error_response(response).await);
        }
        response
            .json()
            .await
            .map_err(|e| SdkError::Protocol(e.to_string()))
    }

    /// `POST /auth/cross-node/deny` (closed out by #741/#749) — the
    /// approver denying a pending cross-node login by its `user_code`.
    /// Deliberately unauthenticated (see
    /// `crates/server/src/cross_node_login.rs::deny`'s own doc comment: a
    /// denial grants nothing, so the worst case of a guessed code is
    /// griefing one's own pending request, not a security bypass).
    pub async fn deny_cross_node_login(
        &self,
        user_code: &str,
    ) -> Result<CrossNodeLoginDenial, SdkError> {
        let response = crate::http::send(&self.http, &self.config.retry, false, |c| {
            c.post(format!("{}/auth/cross-node/deny", self.config.server_url))
                .json(&serde_json::json!({ "user_code": user_code }))
        })
        .await?;
        if !response.status().is_success() {
            return Err(crate::http::map_error_response(response).await);
        }
        response
            .json()
            .await
            .map_err(|e| SdkError::Protocol(e.to_string()))
    }

    /// `GET /identities/{id}/locations` (closed out by #741/#749) —
    /// every location (server base URL) currently advertised for
    /// `identity_id`, the DHT-backed set a requesting node resolves
    /// *before* cross-node login can even start (there is routinely no
    /// session yet at this point — see this method's own unauthenticated
    /// posture, matching `crates/server/src/identity_locator.rs::get_locations`).
    /// Empty, never an error, for an identity with no advertised location
    /// (e.g. no DHT identity configured on the node it's home to).
    pub async fn identity_locations(&self, identity_id: Uuid) -> Result<Vec<String>, SdkError> {
        #[derive(Deserialize)]
        struct LocationsResponse {
            locations: Vec<String>,
        }

        let response = crate::http::send(&self.http, &self.config.retry, true, |c| {
            c.get(format!(
                "{}/identities/{identity_id}/locations",
                self.config.server_url
            ))
        })
        .await?;
        if !response.status().is_success() {
            return Err(crate::http::map_error_response(response).await);
        }
        let body: LocationsResponse = response
            .json()
            .await
            .map_err(|e| SdkError::Protocol(e.to_string()))?;
        Ok(body.locations)
    }
}

impl CrossNodeLogin<'_> {
    /// Drives `POST /auth/cross-node/poll` to completion — identical
    /// backoff shape to `device_login::DeviceLogin::wait`. Resolves to a
    /// real [`Session`] on `approved`, or a typed [`SdkError`] on
    /// `denied`/`expired`.
    pub async fn wait(&self) -> Result<Session, SdkError> {
        let mut interval = self.poll_interval.max(1);

        loop {
            tokio::time::sleep(std::time::Duration::from_secs(interval as u64)).await;

            let response =
                crate::http::send(&self.client.http, &self.client.config.retry, true, |c| {
                    c.post(format!(
                        "{}/auth/cross-node/poll",
                        self.client.config.server_url
                    ))
                    .bearer_auth(&self.request_code)
                })
                .await?;
            if !response.status().is_success() {
                return Err(crate::http::map_error_response(response).await);
            }
            let body: PollCrossNodeLoginResponse = response
                .json()
                .await
                .map_err(|e| SdkError::Protocol(e.to_string()))?;

            match body.status.as_str() {
                "pending" => continue,
                "slow_down" => {
                    interval = (interval * 2).min(MAX_POLL_INTERVAL_SECONDS);
                    continue;
                }
                "denied" => return Err(SdkError::CrossNodeLoginDenied),
                "approved" => {
                    // Single-use delivery, same invariant
                    // `device_login::DeviceLogin::wait`'s own doc comment
                    // documents for the identical reason.
                    let token = body.token.ok_or_else(|| {
                        SdkError::Protocol("approved poll response was missing a token".to_string())
                    })?;
                    return self.client.authenticate(&token).await;
                }
                _ => return Err(SdkError::CrossNodeLoginExpired),
            }
        }
    }
}

#[cfg(test)]
mod approver_tests {
    use super::*;

    #[test]
    fn cross_node_login_lookup_deserializes_from_the_documented_wire_shape() {
        let raw = serde_json::json!({
            "status": "pending",
            "requesting_context": "https://ashen-realms.example",
            "expires_in": 300,
            "integrator_verified": true,
            "display_name": "Ashen Realms",
        });
        let lookup: CrossNodeLoginLookup = serde_json::from_value(raw).unwrap();
        assert_eq!(lookup.status, "pending");
        assert!(lookup.integrator_verified);
        assert_eq!(lookup.display_name.as_deref(), Some("Ashen Realms"));
    }

    #[test]
    fn cross_node_login_denial_deserializes_from_the_documented_wire_shape() {
        let raw = serde_json::json!({ "status": "denied" });
        let denial: CrossNodeLoginDenial = serde_json::from_value(raw).unwrap();
        assert_eq!(denial.status, "denied");
    }
}
